### PR TITLE
added interlocking to calculate active set size

### PR DESCRIPTION
**Pull request by antonlerner**

### CODE REVIEW

File: activation/atxdb.go [hunk 1]
```diff
@@ -192,8 +201,28 @@ func (db *ActivationDb) CalcActiveSetFromView(view []types.BlockID, pubEpoch typ
...
+		db.assLock.Unlock()
+		// if there is a running calculation, wait for it to end and get the result
+		mu.Lock()
+		count, found := activesetCache.Get(viewHash)
+		if found {
+			mu.Unlock()
+			return count, nil
+		}
+		// if not found, keep running mutex and calculate active set size
+	} else {
```
Comment (gavraz) [inline activation/atxdb.go:220]: If not running does not necessarily mean we don't have the calculation in the cache.

File: activation/activationdb_test.go [hunk 2]
```diff
@@ -239,6 +244,49 @@ func TestMesh_ActiveSetForLayerView2(t *testing.T) {
...
+
+	blocks := createLayerWithAtx(t, layers, 1, 10, atxs, []types.BlockID{}, []types.BlockID{})
+	blocks = createLayerWithAtx(t, layers, 10, 10, []*types.ActivationTx{}, blocks, blocks)
+	blocks = createLayerWithAtx(t, layers, 100, 10, []*types.ActivationTx{}, blocks, blocks)
+
+	atx := types.NewActivationTx(id1, coinbase1, 1, atxs[0].Id(), 1000, 0, atxs[0].Id(), 3, blocks, &types.NIPST{})
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	for i := 0; i < 2; i++ {
```
Comment (gavraz) [inline activation/activationdb_test.go]: Should run for many (1000) requests concurrently.
Should calc for more than one single view.

File: activation/activationdb_test.go [hunk 3]
```diff
@@ -239,6 +244,49 @@ func TestMesh_ActiveSetForLayerView2(t *testing.T) {
...
+		go func() {
+			num, err := atxdb.CalcActiveSetFromView(atx.View, atx.PubLayerIdx.GetEpoch(layersPerEpochBig))
+			assert.NoError(t, err)
+			assert.Equal(t, 3, int(num))
```
Comment (gavraz) [inline activation/activationdb_test.go:293]: A good test of concurrency will also use a mock and check the number of calls to each function.
For example, running 1000 calculations for layer x then asserting #calls to the actual calculation = 1 and total #calls to get from cache should be 999.

File: activation/atxdb.go [hunk 4]
```diff
@@ -192,8 +201,28 @@ func (db *ActivationDb) CalcActiveSetFromView(view []types.BlockID, pubEpoch typ
...
+	if alreadyRunning {
+		db.assLock.Unlock()
+		// if there is a running calculation, wait for it to end and get the result
+		mu.Lock()
+		count, found := activesetCache.Get(viewHash)
+		if found {
+			mu.Unlock()
+			return count, nil
+		}
+		// if not found, keep running mutex and calculate active set size
```
Comment (gavraz) [inline activation/atxdb.go:219]: misplaced comment

**Review by gavraz — CHANGES_REQUESTED** [review]

File: activation/activationdb_test.go (hunk 3 above)
Comment (antonlerner) [reply activation/activationdb_test.go:293]: this required heavy refactoring of either the activationDb and it's dependancies

File: activation/atxdb.go (hunk 4 above)
Comment (antonlerner) [reply activation/atxdb.go:219]: this is exactly the place for this comment. this explains the logic when no item in cache and no running calculation

File: activation/atxdb.go (hunk 1 above)
Comment (antonlerner) [reply activation/atxdb.go:220]: this is checked at the beginning of the function

File: activation/activationdb_test.go (hunk 3 above)
Comment (gavraz) [reply activation/activationdb_test.go:293]: תדחיף לו

**Review by gavraz — APPROVED** [review]

**Pull request merged**